flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1375,32 +1375,60 @@ class ScalarExprLowering {
-    // this coordinate op will only work if the extents are compile
-    // time constants. Otherwise, the coordinateOp need to be collapsed.
-    for (auto &subsc : aref.subscript())
-      args.push_back(genunbox(subsc));
-    auto ty = genSubType(base.getType(), args.size());
-    ty = builder.getRefType(ty);
-    auto addr = builder.create<fir::CoordinateOp>(getLoc(), ty, base, args);
-    return arrayElementToExtendedValue(builder, getLoc(), component, addr);
-  }
-
   static bool isSlice(const Fortran::evaluate::ArrayRef &aref) {
     for (auto &sub : aref.subscript())
       if (std::holds_alternative<Fortran::evaluate::Triplet>(sub.u))
         return true;
     return false;
   }
 
-  ExtValue gen(const Fortran::lower::SymbolBox &si,
-               const Fortran::evaluate::ArrayRef &aref) {
+  static bool isDynamicLengthCharacter(const ExtValue &exv) {
+    auto ty = fir::unwrapSequenceType(
+        fir::dyn_cast_ptrOrBoxEleTy(fir::getBase(exv).getType()));
+    return fir::characterWithDynamicLen(ty);
+  }
+
+  /// Lower an ArrayRef to a fir.coordinate_of given its lowered base.
+  ExtValue genCoordinateOp(const ExtValue &array,
+                           const Fortran::evaluate::ArrayRef &aref) {
+    auto loc = getLoc();
+    // References to array of rank > 1 with non constant shape that are not
+    // fir.box must be collapsed into an offset computation in lowering already.
+    // The same is needed with dynamic length characters.
+    if ((array.rank() > 1 && !array.hasConstantShape()) ||
+        isDynamicLengthCharacter(array))
+      if (!array.getBoxOf<fir::BoxValue>())
+        return genCollapsedCoordinateOp(array, aref);
+    // Generate a fir.coordinate_of with zero based array indexes.
+    llvm::SmallVector<mlir::Value> args;
+    for (auto &subsc : llvm::enumerate(aref.subscript())) {
+      auto subVal = genSubscript(subsc.value());
+      assert(fir::isUnboxedValue(subVal) && "subscript must be simple scalar");
+      auto val = fir::getBase(subVal);
+      auto ty = val.getType();
+      auto lb = getLBound(array, subsc.index(), ty);
+      args.push_back(builder.create<mlir::SubIOp>(loc, ty, val, lb));
+    }
+
+    auto base = fir::getBase(array);
+    auto seqTy =
+        fir::dyn_cast_ptrOrBoxEleTy(base.getType()).cast<fir::SequenceType>();
+    assert(args.size() == seqTy.getDimension());
+    auto ty = builder.getRefType(seqTy.getEleTy());
+    auto addr = builder.create<fir::CoordinateOp>(loc, ty, base, args);
+    return arrayElementToExtendedValue(builder, loc, array, addr);
+  }
+
+  /// Lower an ArrayRef to a collapsed fir.coordinate_of given its lowered base.
+  /// This generates offset computation from the indexes and length parameters,
+  /// and use the offset to access the element with a fir.coordinate_of. This
+  /// must only be used if there is not possible to generate a normal
+  /// fir.coordinate_of using array indexes (i.e. when the shape information is
+  /// lost in the IR).
```

Comment:
s/lost/unavailable/

"Lost" makes it sound like something is broken in the IR, when it is just a existential fact that Fortran arrays may have extents that aren't known at compile time, and lowering is doing the correct thing when creating array types with unknown extents.